File: python/Q312_burst_balloons.py
#
# @lc app=leetcode id=312 lang=python3
#
# [312] Burst Balloons
#


# @lc code=start
class Solution:
    def maxCoins(self, nums: list[int]) -> int:
        n: int = len(nums)
        new_nums: list[int] = [1] + nums + [1]

        dp: list[list[int]] = [[0] * (n + 2) for _ in range(n + 2)]

        for left in range(n, 0, -1):
            for right in range(left, n + 1):
                for i in range(left, right + 1):
                    coins: int = new_nums[left - 1] * new_nums[i] * new_nums[right + 1]
                    coins += dp[left][i - 1] + dp[i + 1][right]
                    dp[left][right] = max(dp[left][right], coins)

        return dp[1][n]


# @lc code=end

File: rust/src/leetcode/Q312_burst_balloons.rs
/*
 * @lc app=leetcode id=312 lang=rust
 *
 * [312] Burst Balloons
 */

// @lc code=start
impl Solution {
    pub fn max_coins(nums: Vec<i32>) -> i32 {
        let n: usize = nums.len();

        let mut new_nums: Vec<i32> = vec![1];
        new_nums.extend(nums);
        new_nums.push(1);

        let mut dp: Vec<Vec<i32>> = vec![vec![0; n + 2]; n + 2];

        for l in (1..=n).rev() {
            for r in l..=n {
                for i in l..=r {
                    let coins: i32 = new_nums[l - 1] * new_nums[i] * new_nums[r + 1]
                        + dp[l][i - 1]
                        + dp[i + 1][r];

                    dp[l][r] = dp[l][r].max(coins);
                }
            }
        }

        dp[1][n]
    }
}
// @lc code=end
